IdentityCore/src/cache/MSIDTokenCache.m

```diff
@@ -24,50 +24,386 @@
 #import "MSIDTokenCache.h"
 #import "MSIDToken.h"
 #import "MSIDTokenCacheKey.h"
+#import "MSIDTokenCacheDataSource.h"
+#import "MSIDKeyedArchiverSerializer.h"
+#import "MSIDJsonSerializer.h"
+#import "MSIDUser.h"
 
 @implementation MSIDTokenCache
+{
+    id<MSIDTokenCacheDataSource> _dataSource;
+    MSIDJsonSerializer *_jsonSerializer;
+    MSIDKeyedArchiverSerializer *_keyedArchiverSerialize;
+}
 
-- (BOOL)saveToken:(MSIDToken *)token
-           forKey:(MSIDTokenCacheKey *)key
-            error:(NSError **)error
+- (id)initWithDataSource:(id<MSIDTokenCacheDataSource>)dataSource
 {
-    return NO;
+    if (!(self = [super init]))
+    {
+        return nil;
+    }
+    
+    _dataSource = dataSource;
+    return self;
 }
 
-- (MSIDToken *)getAdalAT:(NSString *)userId
-                resource:(NSString *)resource
-                clientId:(NSString *)clientId
-                   error:(NSError **)error
+- (id<MSIDTokenCacheDataSource>)dataSource
 {
-    return nil;
+    return _dataSource;
 }
 
-- (MSIDToken *)getMsalAT:(NSString *)userId
-                  scopes:(NSSet<NSString *> *)scopes
-                clientId:(NSString *)clientId
-                   error:(NSError **)error
+- (MSIDToken *)getAdalATRTforUser:(MSIDUser *)user
+                        authority:(NSURL *)authority
+                         resource:(NSString *)resource
+                         clientId:(NSString *)clientId
+                          context:(id<MSIDRequestContext>)context
+                            error:(NSError **)error
 {
-    return nil;
+    MSIDTokenCacheKey *key = [MSIDTokenCacheKey keyWithAuthority:authority
+                                                        clientId:clientId
+                                                        resource:resource
+                                                             upn:user.upn];
+    
+    return [_dataSource itemWithKey:key
+                         serializer:_keyedArchiverSerialize
+                            context:context error:error];
 }
 
-- (MSIDToken *)getFRT:(NSString *)userId
-             familyId:(NSString *)familyId
-                error:(NSError **)error
+
+- (MSIDToken *)getFRTforUser:(MSIDUser *)user
+                   authority:(NSURL *)authority
+                    familyId:(NSString *)familyId
+                     context:(id<MSIDRequestContext>)context
+                       error:(NSError **)error
 {
-    return nil;
+    
+    if (!familyId)
+    {
+        familyId = @"1";
+    }
+    NSString *fociClientId = [NSString stringWithFormat:@"foci-%@", familyId];
+    MSIDTokenCacheKey *key = [MSIDTokenCacheKey keyWithAuthority:authority
+                                                        clientId:fociClientId
+                                                        resource:nil
+                                                             upn:user.upn];
```

Comment:
Should we reuse RT saving method here instead and just populate fociClientId?

Reply:
yes, changed.